Implementation encrypt/decrypt feature

logvik opened the pull request:
The new feature, as requested here https://github.com/MetaMask/metamask-extension/issues/1190
Instead https://github.com/MetaMask/metamask-extension/pull/7247

logvik committed:
Implementation for "eth_decryptMessage". It allows using private user keys via keyring and decrypting messages that were encrypted via eth-sig-util.
- Need to add translations for all languages, except en and ru
- Need to fix issue with badges. The decrypt request does not influence on badges, although code for this part is present. The core team, please review this piece.

eth-json-rpc-middleware has updated on 4.3.0

the new eth-keyring-controller version

Added EncryptionPublicKey and decryptMsgInline features

logvik committed:
npm version for eth-json-rpc-middleware

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

# Conflicts:
#	app/scripts/controllers/permissions/permissions-safe-methods.json
#	app/scripts/metamask-controller.js
#	package.json
#	ui/app/store/actions.js
#	ui/lib/tx-helper.js
#	yarn.lock

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
merged with origin repository

logvik committed:
lint

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik commented:
```
error Couldn't find match for "027447b83452a6f8b3e29bfa4b142657786f4d43" in "refs/heads/greenkeeper/initial,refs/heads/master,refs/heads/v3" for "https://github.com/logvik/eth-json-rpc-filters".
```
@Gudahtt  I think that we are forced to publish updated packages on npm before this PR. `yarn.lock` won't help in this case. I can be wrong of course.
P. s. I can run command `yarn --frozen-lockfile --ignore-scripts --har` without any problem on my PC.

Gudahtt committed:
Use published versions of dependencies

The latest published versions of `eth-keyring-controller`,
`eth-json-rpc-infura`, and `eth-json-rpc-filters` are now used. They
in turn are using the latest version of `eth-hd-keyring` and
`eth-json-rpc-middleware` respectively.

Gudahtt commented:
@logvik I've done it here: https://github.com/logvik/metamask-extension/pull/1

logvik committed:
Merge pull request #1 from Gudahtt/use-published-deps

Use published versions of dependencies

logvik committed:
fix for yarn, eth-json-rpc-middleware from NPM

Gudahtt commented on yarn.lock:
```diff
@@ -10436,7 +10436,8 @@ eth-json-rpc-middleware@^1.5.0:
 eth-json-rpc-middleware@^4.1.4, eth-json-rpc-middleware@^4.1.5, eth-json-rpc-middleware@^4.4.0:
   version "4.4.0"
-  resolved "https://github.com/MetaMask/eth-json-rpc-middleware#a0f9d05b5f4e229621a9558c17fa4acb920de822"
+  resolved "https://registry.yarnpkg.com/eth-json-rpc-middleware/-/eth-json-rpc-middleware-4.4.0.tgz#ef63b783b48dcbea9c1fe25c79e6ea01510e5877"
```
Oops, good catch

logvik committed:
fixes for i18n, mock-state, unapproved messages count

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

danfinlay commented on app/scripts/controllers/permissions/permissions-safe-methods.json:
```diff
@@ -46,6 +46,8 @@
   "eth_syncing",
   "eth_uninstallFilter",
+  "encryption_public_key",
+  "eth_decryptMessage",
```
These methods should both probably use the `wallet_` prefix, as they are not eth-related, but they are related to the user's accounts, which is what we have used `wallet_` to mean.

logvik replied:
then we need to rewrite code to another npm package - https://github.com/MetaMask/eth-json-rpc-middleware/blob/master/wallet.js, but I see there 'personal_sign' as 'encryption_public_key ' and  'eth_signTypedData' as 'eth_decryptMessage'. Let's come to the single agreement for global naming and migrate to it. I think that need to do it simultaneously.

danfinlay commented on ui/app/store/actions.js:
```diff
@@ -645,6 +645,89 @@ export function signPersonalMsg (msgData) {
+      background.decryptMessageInline(msgData, (err, newState) => {
+        log.debug('decryptMsg called back')
+        dispatch(updateMetamaskState(newState))
+        msgData = newState.unapprovedDecryptMsgs[msgData.metamaskId]
```
I got an error here: `Error in event handler: TypeError: Cannot read property 'unapprovedDecryptMsgs' of undefined`

logvik replied:
Please provide more information about how I can reproduce this bug. Also, please watch the video about user flow in the test dapp.

danfinlay reviewed (dismissed):
Thanks so much for this enormous contribution. I'm so sorry we've been so slow on this, as you know we're working on a few things to make adding new methods easier in the future.

I noticed some issues here, like wording, but also a few other things:

- The test dapp does not have a method to encrypt a string, so I was unable to test the decryption.
- Copying the encrypted text crashed the UI, so I was unable to do anything in the window after doing it (I noted where a crash seemed to be)
- I'm not sure we need permission to get a public key after the user has already signed in with an account, but we can solve this later.
- We need to have the name of the requesting site on the public key request.

Other than that, it's looking very good. I can see you implemented the styles provided by our design team, and it's come a long way. I think a little bit of wording and stability improvement and this will be good to merge.

logvik committed:
updated version for eth-keyring-controller

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
i18n issues are fixed

logvik commented on app/_locales/en/messages.json:
```diff
@@ -1571,5 +1571,29 @@
+    "message": "Provide"
+  },
+  "encryptionPublicKeyRequest": {
+    "message": "Request ecryption public key"
```
Fixed https://github.com/logvik/metamask-extension/commit/7499aede095383a34d5e4bfb3c99b13879fd0891

logvik commented on app/_locales/en/messages.json:
```diff
@@ -1571,5 +1571,29 @@
+    "message": "Request ecryption public key"
+  },
+  "ecryptionPublickKeyNotice": {
+    "message": "By granted this encryption public key for Dapp you will allow to get encrypted messages and decrypt it via Metamask. This key can be used ONLY for encryption messages that decrypt via your private key. Dapp can't sign transaction or something else via this public key."
```
Fixed https://github.com/logvik/metamask-extension/commit/7499aede095383a34d5e4bfb3c99b13879fd0891

logvik commented on app/_locales/en/messages.json:
```diff
@@ -1571,5 +1571,29 @@
+  "encryptionPublicKeyRequest": {
+    "message": "Request ecryption public key"
+  },
+  "ecryptionPublickKeyNotice": {
```
Fixed https://github.com/logvik/metamask-extension/commit/7499aede095383a34d5e4bfb3c99b13879fd0891

logvik committed:
the fix in wording

logvik commented:
Thank a lot for reviewing.

- The test dapp does not have a method to encrypt a string, so I was unable to test the decryption.

**_There has the field for the message input. I have recorded the video, please watch it._**

- Copying the encrypted text crashed the UI, so I was unable to do anything in the window after doing it (I noted where a crash seemed to be)

**_I have recorded the video, please watch it._**

- I'm not sure we need permission to get a public key after the user has already signed in with an account, but we can solve this later.

**_We need it because with `encryption_public_key` method the user expresses agreement to receive encrypted messages._**

- We need to have the name of the requesting site on the public key request.

**_I think we have it. But if we don't have it on the recorded video, please provide more information about this moment._**

![metamask](https://user-images.githubusercontent.com/2010097/72836142-1c446d80-3c9d-11ea-8124-02201d40147b.gif)

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
fixed quotes for lint

MartyHav commented:
> * The test dapp does not have a method to encrypt a string, so I was unable to test the decryption.
@danfinlay 

Just to avoid a possible misunderstanding: the message is encrypted on any 'keyup' next to the
input field. Looks only static if its not expanded.

tmashuang commented:
There are a couple of user facing edge cases that I hit when trying this out on the `test-dapp` branch with the encryption/decryption functions. 

- Clicking `Decrypt message` before providing the `encryptionpublickey` will show the notification screen and clicking Decrypt on the ConfirmDecryptMessage Component will result in an indefinite loading screen. 

  - Would it be possible to prevent this behavior by checking if the window has been provided an `encryptionpublickey`?

- While going through the correct flow, the `Copy encrypted message`  will not be copied before the message has been shown/unlocked. 

  - Possible disable the button click, or show a message for the tooltip to copy after showing/unlocking the message.

Obviously these are edge cases and don't detract from the extensive work already put into this feature. The majority of the functionality has been added and am happy to merge as is with the request that another dev, or two, and possibly a designer to sign off on this with the capacity to solve these minor details.

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
1) fixed bug when the message for decryption is empty
2) coping decrypted message is available only after inline decryption
3) name of the requesting site on the public key request

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik commented:
@tmashuang 
> * Clicking `Decrypt message` before providing the `encryptionpublickey` will show the notification screen and clicking Decrypt on the ConfirmDecryptMessage Component will result in an indefinite loading screen.
>   
>   * Would it be possible to prevent this behavior by checking if the window has been provided an `encryptionpublickey`?

It is fixed for now. The problem has been in the empty message.

> * While going through the correct flow, the `Copy encrypted message`  will not be copied before the message has been shown/unlocked.
>   
>   * Possible disable the button click, or show a message for the tooltip to copy after showing/unlocking the message.

It is fixed for now. This button is hidden before the user will see the message.

@danfinlay 
> * We need to have the name of the requesting site on the public key request.

It is fixed for now.

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

Gudahtt commented on app/scripts/controllers/threebox.js:
```diff
@@ -50,6 +50,16 @@ class ThreeBoxController {
           withAppKeyOrigin: 'wallet://3box.metamask.io',
         })
       },
+      processDecryptMessage: (msgParams) => {
```
I don't think these two functions are required. This 3Box-specific provider is solely used to facilitate 3Box validating that the box they're creating is authorized by the user - it doesn't use these methods. That's why the `processTypedMessage` functions aren't implemented either, for example.

logvik replied:
I have removed it. Thank you

logvik committed:
removed unnecessary methods for 3box

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

whymarrh commented:
@logvik I've factored out the changes to the `package.json` and `yarn.lock` files—please drop those changes from this PR (cross-linked above)

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik commented:
@whymarrh , I have done merge from the origin repository. Will it be enough or should I replace content in package.js and yarn.lock from the origin wholly?

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+  * Called when a dapp uses the eth_decryptMessage method.
+  *
+  * @param {Object} msgParams - The params of the message to sign & return to the Dapp.
+  * @param {Function} cb - The callback function called with the signature.
```
This is inaccurate - this should document the `req` parameter.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+  /**
+  * Only decypt message and don't touch transaction state
+  *
+  * @param {Object} msgParams - The params of the message to decrypt & return to the Dapp.
```
```suggestion
  * @param {Object} msgParams - The params of the message to decrypt
```
I'd suggest removing that last part of the description - not that it's inaccurate technically, but it implied to me when I first read it that this function returned the decrypted message to the dapp.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.decryptMessageManager.approveMessage(msgParams)
```
Could you update this to use `async/await` instead? This is a good example of why traditional Promises can be error-prone - if an error occurs in this chain, it does get returned but it won't be caught by the `catch` block.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+        })
+    } catch (error) {
+      log.info('MetaMaskController - eth_decryptMessage failed.', error)
+      this.decryptMessageManager.errorMessage(msgId, error)
```
This should probably re-throw the error, so that the UI knows something wrong happened.

whymarrh commented:
> Will it be enough or should I replace content in package.js and yarn.lock from the origin wholly?

The latter should work: `git fetch origin && git checkout origin/develop -- package.json yarn.lock`

The changes in those PRs don't match 1:1 to the changes you have here so merging doesn't quite do the job alone.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,148 @@ export default class MetamaskController extends EventEmitter {
+    // sets the status op the message to 'approved'
+    // and removes the metamaskId for decryption
+    try {
+      return this.encryptionPublicKeyManager.approveMessage(msgParams)
```
The same two comments apply here as well - this should use `async/await`, and the catch block should re-throw the error.

whymarrh commented:
Feel free to use merges to update as well, if that's what you're most comfortable with

Gudahtt reviewed (changes requested):
I'm finished reviewing the background changes - there are a couple of things that should change, but mostly this is pretty solid. Thanks for this - I appreciate the effort you've taken to match the style elsewhere in the codebase.

I'll review the UI shortly and make specific suggestions on how it could be improved, but to start with, here are the latest designs for the decrypt message popup: https://www.figma.com/file/7QEoPXhxxq2sfBNZWT2OX9/Encryption?node-id=133%3A770
There are a few notable differences from the one implemented here, which I'm assuming was based upon the earlier mockups that were shared. There is no design for the public key popup, but we'll want that one to look similar as well (at least for the header and footer)

logvik committed:
package.json and yarn.lock are given by the origin

logvik committed:
fixes after Gudahtt reviewing

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik commented:
@whymarrh I have done that you have been asked to do me with package.js and yarn.lock. Please review.

@Gudahtt I have fixed remarks that you have been mentioned after reviewing the background changes.

Gudahtt reviewed (changes requested):
I've reviewed part of the UI - I just have the two components left. I need to refresh my memory on how the component hierarchy works for our confirmation screens, then I'll get back to you on those.

Gudahtt commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,330 @@
+  render = () => {
+    return (
+      <div className="request-decrypt-message__container">
+        { this.renderHeader() }
```
So the key difference between this current version and the most recent designs is the use of the `ConfirmTransactionBase` wrapper component (and the header and footer that are used by that). Removing this header and footer, and passing in `renderBody()` to `ConfirmTransactionBase` as the `contentComponent` prop should make the header and footer conform to the designs. Take a look at the `ui/app/pages/confirm-approve/confirm-approve.component.js` for an example of how to do this.

The same goes for `confirm-encryption-public-key.component.js` as well - it should use `ConfirmTransactionBase` instead of these custom headers/footers.

logvik replied:
We can't use `ConfirmTransactionBase`, because we don't need to have a gas estimator and other components for such dialog. Also such component very similar on `personal_sign` by design. I don't sure that we need to have different designs for similar operations.

Gudahtt replied:
Ah I see - I hadn't realized that `ConfirmTransactionBase` included transaction costs and gas estimations. You're right, that doesn't make sense. Hmm.

Gudahtt commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,330 @@
+
+    return (
+      <div className="request-decrypt-message__body">
+        { this.renderAccountInfo() }
```
After the switch to using `ConfirmTransactionBase`, this account information will be in the header, so it'll no longer be needed here.

(this also applies to `confirm-encryption-public-key.component.js`)

logvik replied:
Please, look at my last comment about this case.

Gudahtt commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,330 @@
+    const { txData } = this.props
+
+    const origin = this.props.domainMetadata[txData.msgParams.origin]
+    const notice = this.context.t('decryptMessageNotice', ['[' + origin.name + ']'])
```
```suggestion
    const notice = this.context.t('decryptMessageNotice', [origin.name)
```
These extra square brackets are not needed - the square brackets in the mockups/designs were just meant to indicate that the contents were variable.

Gudahtt commented on app/_locales/en/messages.json:
```diff
@@ -1603,5 +1603,31 @@
+	"description": "$1 is website or dapp name"
+  },
+  "decryptMetamask": {
+    "message": "Decrypt with Metamask"
```
```suggestion
    "message": "Decrypt message"
```
The designs have this as "Decrypt message", which I think works a bit better. The `ru` message will need to be updated as well.

Gudahtt commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,330 @@
+                title={hasCopied ? this.context.t('copiedExclamation') : this.context.t('copyToClipboard')}
+                wrapperClassName="request-decrypt-message__message-copy-tooltip"
+              >
+                <div
```
This "copy to clipboard" part should be centred horizontally, and should be this color: `#037DD6`

Gudahtt reviewed (changes requested):
Alright, all done! The switch to `ConfirmTransactionBase` is the biggest change needed for the UI.

Sorry for all the comments :sweat_smile: This is quite a substantial feature. Please feel free to ask us to take over if you're short on time or enthusiasm; we'll make sure the work is completed.

logvik committed:
fix for the last review by Gudahtt

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik commented on ui/app/store/actions.js:
```diff
@@ -645,6 +645,95 @@ export function signPersonalMsg (msgData) {
+          msgData = newState.unapprovedDecryptMsgs[msgData.metamaskId]
+          dispatch(updateMetamaskState(newState))
+        } else {
+          log.error('Wrong data for decryptMessageInline')
```
We need to have this for a case when the message for decryption is empty.

Gudahtt replied:
I'm not sure I follow - what happens when the message for decryption is empty? I would expect it to either throw an error in that case, or return a full state object regardless. I can't see a path where it'd something falsey in place of this state object _and_ not throw an error.

Gudahtt replied:
I've just tested it, and this has the effect of catching _all errors_ and returning a "Wrong data" message, rather than the error itself. We should be rejecting the actual `Error` object when an error occurs.

logvik commented:
I have compared designs for `personal_sign` operation and for `encryption_public_key` with `eth_decryptMessage`. All are the same. For this reason, I don't see any benefits to use new design for a popup that offered by this pull request. Please correct me if  I am wrong.
![personal_sign](https://user-images.githubusercontent.com/2010097/73873329-767b2c00-4862-11ea-85d8-879bb71cae98.gif)

The quote of comment with popups for this pull request:

> Thank a lot for reviewing.
> 
> * The test dapp does not have a method to encrypt a string, so I was unable to test the decryption.
> 
> **_There has the field for the message input. I have recorded the video, please watch it._**
> 
> * Copying the encrypted text crashed the UI, so I was unable to do anything in the window after doing it (I noted where a crash seemed to be)
> 
> **_I have recorded the video, please watch it._**
> 
> * I'm not sure we need permission to get a public key after the user has already signed in with an account, but we can solve this later.
> 
> **_We need it because with `encryption_public_key` method the user expresses agreement to receive encrypted messages._**
> 
> * We need to have the name of the requesting site on the public key request.
> 
> **_I think we have it. But if we don't have it on the recorded video, please provide more information about this moment._**
> 
> ![metamask](https://user-images.githubusercontent.com/2010097/72836142-1c446d80-3c9d-11ea-8124-02201d40147b.gif)

logvik committed:
fix for lint errors

Gudahtt commented:
>I don't see any benefits to use new design for a popup that offered by this pull request

Well, the benefit is that we think the new design is better 😅 . But it probably should match `personal_sign` - that's a fair point. We still want the new designs to be implemented, but perhaps we can do that separately, and update `personal_sign` at the same time 🤔 . I'll see what others think.

Gudahtt commented:
We've decided to accept this current design as-is for now, and we can implement the newer designs in a separate PR. There are also updated designs for the signature request screen - they probably won't be ready to work on right away, but hopefully soon.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,143 @@ export default class MetamaskController extends EventEmitter {
+    const msgId = msgParams.metamaskId
+    const stripped = ethUtil.stripHexPrefix(msgParams.data)
+    const buff = Buffer.from(stripped, 'hex')
+    msgParams.data = JSON.parse(buff.toString('utf8'))
```
Here and on line 1242, the JSON.parse blows up if the message is empty. This seems like the right place to check for an empty message, and return immediately without throwing an error (e.g. "decrypt" it by setting it to an empty string).

Allowing the user to 'decrypt' an empty string is odd, but it seems a bit more sensible than disallowing these messages from being added in the first place? Or at it at least seems easier.

logvik replied:
fixed

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1168,6 +1185,143 @@ export default class MetamaskController extends EventEmitter {
+    const msgId = msgParams.metamaskId
+    const stripped = ethUtil.stripHexPrefix(msgParams.data)
+    const buff = Buffer.from(stripped, 'hex')
+    msgParams.data = JSON.parse(buff.toString('utf8'))
```
Also, an error thrown here beaks our error handling - nothing gets returned to the caller, so the error is lost.

Switching this to use `async/await` would be the best way of ensuring all errors are propagated correctly.

logvik replied:
fixed

logvik committed:
better handling for edges cases

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

Gudahtt reviewed (dismissed):
There are still some issues with error handling (from the user's perspective it still seems to "silently fail" now in some cases), but it's about on-bar with the signature request UI, and it'd be challenging to improve it a great deal as we don't have any designs or UI patterns to fall back on for unrecoverable confirmation screen errors. I'll create a separate issue for tracking that problem.

Great work @logvik!

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+  _beforeUnload = (event) => {
+    const { clearConfirmTransaction, cancelEncryptionPublicKey } = this.props
+    const { metricsEvent } = this.context
+    metricsEvent({
```
@Gudahtt we should provide guidelines for good metrics events, will follow up on this and provide the requested change

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+          onClick={async event => {
+            this._removeBeforeUnload()
+            await this.props.cancelEncryptionPublicKey(txData, event)
+            this.context.metricsEvent({
```
@Gudahtt metrics copy note, will follow up with revised message

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+          onClick={async event => {
+            this._removeBeforeUnload()
+            await this.props.encryptionPublicKey(txData, event)
+            this.context.metricsEvent({
```
@Gudahtt metricsEvent

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+  _beforeUnload = (event) => {
+    const { clearConfirmTransaction, cancelDecryptMessage } = this.props
+    const { metricsEvent } = this.context
+    metricsEvent({
```
@Gudahtt metricsEvent

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+
+  copyMessage = () => {
+    copyToClipboard(this.state.rawMessage)
+    this.context.metricsEvent({
```
@Gudahtt metricsEvent

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+          onClick={async event => {
+            this._removeBeforeUnload()
+            await this.props.cancelDecryptMessage(txData, event)
+            this.context.metricsEvent({
```
@Gudahtt metricsEvent

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+          onClick={async event => {
+            this._removeBeforeUnload()
+            await this.props.decryptMessage(txData, event)
+            this.context.metricsEvent({
```
@Gudahtt metricsEvent

jennypollack commented on ui/index.js:
```diff
@@ -59,7 +59,7 @@ async function startApp (metamaskState, backgroundConnection, opts) {
 
   // if unconfirmed txs, start on txConf page
-  const unapprovedTxsAll = txHelper(metamaskState.unapprovedTxs, metamaskState.unapprovedMsgs, metamaskState.unapprovedPersonalMsgs, metamaskState.unapprovedTypedMessages, metamaskState.network)
+  const unapprovedTxsAll = txHelper(metamaskState.unapprovedTxs, metamaskState.unapprovedMsgs, metamaskState.unapprovedPersonalMsgs, metamaskState.unapprovedDecryptMsgs, metamaskState.unapprovedEncryptionPublicKeyMsgs, metamaskState.unapprovedTypedMessages, metamaskState.network)
```
will these messages be network dependent?

logvik replied:
No, they are not about transactions. But what do you mean? If I will change the network on `testnet`, then will it count by another handler? I didn't check how it works right now, but if each network has own global counter over all requests, then my changes will inherit the same behavior.

Gudahtt replied:
This `txHelper` function only counts the unapproved transactions for the current network - that's why the network is passed in. The changes made here aren't affected by the network. The same goes for the other types of signing.

logvik committed:
fixes after jennypollack review

logvik commented on test/data/2-state.json:
```diff
@@ -18,6 +18,8 @@
   "unapprovedMsgCount": 0,
   "unapprovedPersonalMsgs": {},
   "unapprovedPersonalMsgCount": 0,
+  "unapprovedDecryptMsgs": 0,
```
Fixed

logvik commented on app/scripts/lib/decrypt-message-manager.js:
```diff
@@ -0,0 +1,311 @@
+
+  /**
+   * Approves a DecryptMessage. Sets the message status via a call to this.setMsgStatusApproved, and returns a promise
+   * with any the message params modified for proper decryption.
```
Fixed

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+    const { metricsEvent } = this.context
+    metricsEvent({
+      eventOpts: {
+        category: 'Transactions',
```
```suggestion
        category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+      eventOpts: {
+        category: 'Transactions',
+        action: 'Decrypt Message Request',
+        name: 'Cancel Decrypt Message Request Via Notification Close',
```
```suggestion
        name: 'Cancel Via Notification Close',
```

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+    copyToClipboard(this.state.rawMessage)
+    this.context.metricsEvent({
+      eventOpts: {
+        category: 'Transactions',
```
```suggestion
        category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+            await this.props.cancelDecryptMessage(txData, event)
+            this.context.metricsEvent({
+              eventOpts: {
+                category: 'Transactions',
```
```suggestion
                category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-decrypt-message/confirm-decrypt-message.component.js:
```diff
@@ -0,0 +1,333 @@
+            await this.props.decryptMessage(txData, event)
+            this.context.metricsEvent({
+              eventOpts: {
+                category: 'Transactions',
```
```suggestion
                category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+    const { metricsEvent } = this.context
+    metricsEvent({
+      eventOpts: {
+        category: 'Transactions',
```
```suggestion
        category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+      eventOpts: {
+        category: 'Transactions',
+        action: 'Encryption public key Request',
+        name: 'Cancel Encryption public key Request Via Notification Close',
```
```suggestion
        name: 'Cancel Via Notification Close',
```

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+            await this.props.cancelEncryptionPublicKey(txData, event)
+            this.context.metricsEvent({
+              eventOpts: {
+                category: 'Transactions',
```
```suggestion
                category: 'Messages',
```

jennypollack commented on ui/app/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:
```diff
@@ -0,0 +1,238 @@
+            await this.props.encryptionPublicKey(txData, event)
+            this.context.metricsEvent({
+              eventOpts: {
+                category: 'Transactions',
```
```suggestion
                category: 'Messages',
```

logvik committed:
Fixes for the metric messages

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

rekmarks reviewed (changes requested):
@logvik, first of all, nice work! 

There's just one more hurdle to clear before we can merge this! We need to standardize the method names. You've added the following methods:
```
  'encryption_public_key',
  'eth_decryptMessage',
```
`encryption_public_key` does not fit any Ethereum RPC naming scheme. In order to use the `eth_` prefix, `eth_decryptMessage` should really be in an EIP. The only prior art we could find is [EIP 1024](https://github.com/ethereum/EIPs/pull/1098), by @topealabi.  Per that EIP, I suggest the following method names:

```
eth_getEncryptionPublicKey
eth_decrypt
```

For completeness, I'll note that this feature also implies the existence of the following method, which does not need to be implemented as an RPC method at this time:
```
eth_encrypt
```

Does that sound good to you?

logvik committed:
fixes for names of decrypt/encrypt RPC methods

logvik commented:
@rekmarks , thank you for your review.
About `eth_encrypt` - I think that it needs to be implemented on web3 side, not in Metamask extension. I have implemented something like this in test-dapp https://github.com/logvik/test-dapp/blob/master/src/index.js#L226
I have done https://github.com/MetaMask/eth-json-rpc-middleware/pull/21
The last commit has fixes for naming, but it related by my own repository for `eth-json-rpc-middleware` in package.json . FOR TESTING PURPOSE, NOT FOR MERGING.

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
removed the link on the private npm package

logvik commented:
The last commit has all fixes for demands on the review step.

logvik committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into develop

logvik committed:
fix for lint

rekmarks reviewed (approved):
LGTM!

zulhfreelancer commented:
@logvik thank you for this PR. Do you have any docs/tutorial for developers reference?

Gudahtt commented:
@zulhfreelancer Unfortunately this has not been documented yet, but there is an example of this being used here: https://github.com/logvik/test-dapp

Specifically, [here is where it gets the public key](https://github.com/logvik/test-dapp/blob/master/src/index.js#L212), and [here is where it decrypts the message](https://github.com/logvik/test-dapp/blob/master/src/index.js#L233).

logvik commented:
@zulhfreelancer , @Gudahtt please assist me in finding an example of such documentation and I will provide a description in the same style.

Gudahtt commented:
@logvik The best place to document this would be here: https://metamask.github.io/metamask-docs/API_Reference/Experimental_APIs
Here is the repo for those docs: https://github.com/MetaMask/metamask-docs/blob/master/03_API_Reference/03_Experimental_APIs.md

rekmarks commented:
Hello @logvik, have you had a chance to get started on the documentation so that people can learn how to use this feature?

The URL of the documentation has changed: https://github.com/MetaMask/metamask-docs/blob/master/docs/guide/experimental-apis.md

I recommend that you create a new level 2 section, e.g. ``## Experimental `eth_` methods``, and then list the methods as level 3 sections under that, e.g. `eth_decrypt`

rstormsf commented:
@logvik can u share your testing UI that you have on your video?

rekmarks commented:
@rstormsf it's the default branch of: https://github.com/logvik/test-dapp

That's a fork of an older version of our test dapp. I'll ping here when we've added encrypt/decrypt to our hosted version: https://metamask.github.io/test-dapp/
